### DISCUSSION TITLE
Is there a reason why some methods take ownership instead of just borrowing?

### DISCUSSION THREAD

**Discussion by izagawd** [serenity-rs/serenity#3007] [Show and tell]
for example 
```rust
#[cfg(feature = "model")]
impl From<User> for CreateEmbedAuthor {
    fn from(user: User) -> Self {
        let avatar_icon = user.face();
        Self::new(user.name).icon_url(avatar_icon)
    }
}
```
i know that u can just clone the User, but,

```rust
pub struct User {
    pub id: UserId,
    pub name: String,
    pub avatar: Option<ImageHash>,
    pub bot: bool,
    pub system: bool,
    pub mfa_enabled: bool,
    pub banner: Option<ImageHash>,
    pub accent_colour: Option<Colour>,
    pub locale: Option<String>,
    pub verified: Option<bool>,
    pub email: Option<String>,
    #[serde(default)]
    pub flags: UserPublicFlags,
    #[serde(default)]
    pub premium_type: PremiumType,
    pub public_flags: Option<UserPublicFlags>,
    pub member: Option<Box<PartialMember>>,
}
```
the User struct has a decent chunk of data, so thats a decent chunk of data to clone

is there  a reason for it to pass ownerships instead of borrowing?

**Comment by jamesbt365** [comment]
> is there a reason for it to pass ownerships instead of borrowing?

Because the [From](https://doc.rust-lang.org/std/convert/trait.From.html) trait consumes the value, it does not provide an implementation that takes by reference.

**Reply by izagawd** [reply]
ohh my bad!

**Reply by KiloOscarSix** [reply]
Correct me if I'm mistaken, but I'm fairly sure you can do `From<&User>` and use `fn from(user: &User)`
However you'd probably need to clone `user.name` as it's a String

**Reply by jamesbt365** [reply]
> Correct me if I'm mistaken, but I'm fairly sure you can do `From<&User>` and use `fn from(user: &User)` However you'd probably need to clone `user.name` as it's a String

No. See the docs from the linked trait.

**Reply by jamesbt365** [reply]
Huh, I guess I'm wrong, the docs explicitly states that it *consumes* the value, and it doesn't provide anything for references. But it *does* work.

**Reply by KiloOscarSix** [reply]
Good to know, I've just been doing some further reading on the trait as well, as I've used the reference-From without considering the documentation.
